salt/_states/docker_registry.py

```diff
@@ -0,0 +1,185 @@
+
+    image_manifest = __salt__['docker_registry.get_image_manifest'](
+        name, tls_verify=tls_verify)
+
+    if not archive_path:
+        return _image_managed_no_archive(name, image_manifest, digest)
+
+    if image_manifest is not None and digest is not None:
+        digest_match = image_manifest['Digest'] == digest
+
+        if digest_match:
+            return _image_managed_result(name, True, 'Image up-do-date')
+        elif not digest_match and not force:
+            return _image_managed_result(name, False, 'Image digest conflict')
+        elif not digest_match and force:
+            if __opts__['test']:
+                return _image_managed_result(
+                    name, None, 'Will force-push archive',
+                    pchanges=_image_managed_changes(
+                        name, digest, image_manifest['Diges']
+                    ),
+                )
+            else:
+                result = __salt__['docker_registry.push_docker_archive'](
+                    archive_path=archive_path,
+                    destination=name,
+                    dest_tls_verify=tls_verify)
+
+                if result:
+                    return _image_managed_result(
+                        name, True, 'Force-pushed archive',
+                        changes=_image_managed_changes(
+                            name, digest, image_manifest['Digest']
+                        ),
+                    )
+                else:
+                    return _image_managed_result(
+                        name, False, 'Failed to push image',
+                    )
+        else:
+            assert False, 'Logic error'
+    elif image_manifest is not None and digest is None:
+        if not force:
+            return _image_managed_result(name, True, 'Image exists')
+        else:
+            if __opts__['test']:
+                return _image_managed_result(
+                    name, None, 'Would force-push image',
+                    pchanges=_image_managed_changes(
+                        name, digest, image_manifest['Digest'],
+                    ),
+                )
+            else:
+                result = __salt__['docker_registry.push_docker_archive'](
+                    archive_path=archive_path,
+                    destination=name,
+                    dest_tls_verify=tls_verify)
+                if result:
+                    return _image_managed_result(
+                        name, True, 'Force-pushed image',
+                        changes=_image_managed_changes(
+                            name, digest, image_manifest['Digest'],
+                        ),
+                    )
+                else:
+                    return _image_managed_result(
+                        name, False, 'Failed to push image',
+                    )
+
+    elif image_manifest is None:
+        return _image_managed_new_image(name, archive_path, digest, tls_verify)
+    else:
+        assert False, 'Logic error'
```

Comment:
There is probably some code to be merged in this file to ease the reading.

Reply:
I cleaned up quite a bit already after earlier review. I agree, though would like to get the functionality in and rework code if new functionality is required.